metaflow/client/core.py

```diff
@@ -1118,11 +1119,227 @@ class Task(MetaflowObject):
+                field_value = metadata_dict.get("foreach-indices")
+            else:
+                # This is a foreach split
+                # Current Task: foreach-indices = [0, 1, 2], foreach-indices-truncated = [0, 1]
+                # Ancestor Task: foreach-indices = [0, 1], foreach-indices-truncated = [0]
+                # We will compare the foreach-indices value of ancestor task with the
+                # foreach-indices-truncated value of current task
+                field_name = "foreach-indices"
+                field_value = metadata_dict.get("foreach-indices-truncated")
+        else:
+            if query_foreach_stack_len > cur_foreach_stack_len:
+                # This is a foreach split
+                # Current Task: foreach-indices = [0, 1], foreach-indices-truncated = [0]
+                # Successor Task: foreach-indices = [0, 1, 2], foreach-indices-truncated = [0, 1]
+                # We will compare the foreach-indices value of current task with the
+                # foreach-indices-truncated value of successor tasks
+                field_name = "foreach-indices-truncated"
+                field_value = metadata_dict.get("foreach-indices")
+            else:
+                # This is a foreach join
+                # Current Task: foreach-indices = [0, 1, 2], foreach-indices-truncated = [0, 1]
+                # Successor Task: foreach-indices = [0, 1], foreach-indices-truncated = [0]
+                # We will compare the foreach-indices-truncated value of current task with the
+                # foreach-indices value of successor tasks
+                field_name = "foreach-indices"
+                field_value = metadata_dict.get("foreach-indices-truncated")
+        return field_name, field_value
+
+    def _get_related_tasks(self, is_ancestor: bool) -> Dict[str, List[str]]:
+        flow_id, run_id, _, _ = self.path_components
+        metadata_dict = self.metadata_dict
+        steps = (
+            metadata_dict.get("previous-steps")
+            if is_ancestor
+            else metadata_dict.get("successor-steps")
+        )
+
+        if not steps:
+            return {}
+
+        # Convert steps to a list if it's stored as a string in the metadata
+        if is_stringish(steps):
+            steps = [steps]
+
+        field_name, field_value = self._get_metadata_query_vals(
+            flow_id,
+            run_id,
+            len(metadata_dict.get("foreach-indices", [])),
+            steps,
+            is_ancestor=is_ancestor,
+            metadata_dict=metadata_dict,
+        )
+
+        return {
+            step: [
+                f"{flow_id}/{run_id}/{step}/{task_id}"
+                for task_id in self._metaflow.metadata.filter_tasks_by_metadata(
+                    flow_id, run_id, step, field_name, field_value
+                )
+            ]
+            for step in steps
+        }
+
+    @property
+    def immediate_ancestors(self) -> Dict[str, List[str]]:
```

Comment:
can we offer a property `parents` which simply returns a list of task pathspecs. it will return `None`for the `start` step. an open question is if we would want to also offer `parent` - maybe we can cross that bridge as a follow up PR if needed.

Reply:
what are the consistency guarantees offered by this property. do we expect that this property will return an immutable set of parents as soon as this task is registered?

Reply:
also - we are assuming that return values here point to the latest successful attempt. might be good to note this in the doc string.